File: python-version/LC498/find_diagonal_order.py
from typing import List


class Solution:
    @staticmethod
    def findDiagonalOrder(mat: List[List[int]]) -> List[int]:
        row, col = len(mat), len(mat[0])
        cnt = row + col - 1
        m, n, idx = 0, 0, 0
        res = [0] * (row * col)
        for i in range(cnt):
            if i & 1 == 0:
                while m >= 0 and n < col:
                    res[idx] = mat[m][n]
                    idx += 1
                    m -= 1
                    n += 1
                if n < col:
                    m += 1
                else:
                    m += 2
                    n -= 1
            else:
                while m < row and n >= 0:
                    res[idx] = mat[m][n]
                    idx += 1
                    m += 1
                    n -= 1
                if m < row:
                    n += 1
                else:
                    n += 2
                    m -= 1
        return res

File: go-version/leetcode/LC498/find_diagonal_order.go
package LC498

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/8/25 上午7:15
 * @FilePath: find_diagonal_order
 * @Description: lc498，对角线遍历，模拟，注意边界
 */

func findDiagonalOrder(mat [][]int) []int {
	row, col := len(mat), len(mat[0])
	cnt := row + col - 1
	var idx, m, n int
	res := make([]int, row*col)
	for i := 0; i < cnt; i++ {
		if i&1 == 0 { // 偶数个对角线向右上方
			for m >= 0 && n < col { // 没越界正常遍历
				res[idx] = mat[m][n]
				idx++
				m--
				n++
			}
			if n < col { // 处理行越界，直接移动到下一行
				m++
			} else { // 列越界，下移两行，前移一列
				m += 2
				n--
			}
		} else { // 奇数个对角线向左下方
			for m < row && n >= 0 { // 没越界正常遍历
				res[idx] = mat[m][n]
				idx++
				m++
				n--
			}
			if m < row {
				n++
			} else {
				n += 2
				m--
			}
		}
	}
	return res
}
